src/System.Text.Json/src/System/Text/Json/JsonDocument.Parse.cs

```diff
@@ -0,0 +1,276 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Buffers;
+using System.Diagnostics;
+using System.IO;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Text.Json
+{
+    public sealed partial class JsonDocument
+    {
+        private const int UnseekableStreamInitialRentSize = 4096;
+
+        public static JsonDocument Parse(ReadOnlyMemory<byte> utf8Json, JsonReaderOptions readerOptions = default)
+        {
+            CheckSupportedOptions(readerOptions);
+
+            return Parse(utf8Json, readerOptions, null);
+        }
+
+        public static JsonDocument Parse(ReadOnlySequence<byte> utf8Json, JsonReaderOptions readerOptions = default)
+        {
+            CheckSupportedOptions(readerOptions);
+
+            if (utf8Json.IsSingleSegment)
+            {
+                return Parse(utf8Json.First, readerOptions, null);
+            }
+
+            int length = checked((int)utf8Json.Length);
+            byte[] utf8Bytes = ArrayPool<byte>.Shared.Rent(length);
+
+            try
+            {
+                utf8Json.CopyTo(utf8Bytes.AsSpan());
+                return Parse(utf8Bytes.AsMemory(0, length), readerOptions, utf8Bytes);
+            }
+            catch
+            {
+                utf8Bytes.AsSpan(0, length).Clear();
+                ArrayPool<byte>.Shared.Return(utf8Bytes);
+                throw;
+            }
+        }
+
+        public static JsonDocument Parse(Stream utf8Json, JsonReaderOptions readerOptions = default)
+        {
+            if (utf8Json == null)
+                throw new ArgumentNullException(nameof(utf8Json));
+
+            CheckSupportedOptions(readerOptions);
+
+            ArraySegment<byte> drained = ReadToEnd(utf8Json);
+
+            try
+            {
+                return Parse(drained.AsMemory(), readerOptions, drained.Array);
+            }
+            catch
+            {
+                drained.AsSpan().Clear();
+                ArrayPool<byte>.Shared.Return(drained.Array);
+                throw;
+            }
+        }
+
+        public static Task<JsonDocument> ParseAsync(
+            Stream utf8Json,
+            JsonReaderOptions readerOptions = default,
+            CancellationToken cancellationToken = default)
+        {
+            if (utf8Json == null)
+                throw new ArgumentNullException(nameof(utf8Json));
+
+            CheckSupportedOptions(readerOptions);
+
+            return ParseAsyncCore(utf8Json, readerOptions, cancellationToken);
+        }
+
+        private static async Task<JsonDocument> ParseAsyncCore(
```

Comment:
It was a recommendation from @stephentoub to make the argument validation be synchronous code.  One of the tests actually checks for that.

Making the public method async yields:

```
     System.Text.Json.Tests.JsonDocumentTests.ParseNull [FAIL]
       System.InvalidOperationException : You must call Assert.ThrowsAsync, Assert.DoesNotThrowAsync, or Record.ExceptionAsync when testing async code.
       Stack Trace:
         D:\git\bartonjs\corefx\src\CoreFx.Private.TestUtilities\src\System\AssertExtensions.cs(83,0): at System.AssertExtensions.Throws[T](String paramName, Func`1 testCode)
         D:\git\bartonjs\corefx\src\System.Text.Json\tests\JsonDocumentTests.cs(1695,0): at System.Text.Json.Tests.JsonDocumentTests.ParseNull()
```

Which means "the task was successfully returned", vs "the method threw".

Reply:
So it's really for an internal\xunit thing? Since the code is not async until the await.

Reply:
No, it's not just for testing. It's a question of how exceptions propagate. In a method marked async, all exceptions are wrapped in the returned task, regardless of whether they occur before or after an await. But usage exceptions / argument validation should propagate synchronously out of the call.

Reply:
> Since the code is not async until the await.

That depends on what you mean.  If the method has the async keyword then the compiler applies a bunch of transformations, and the first line if code isn't run until the task starts.  This split makes the argument validation happen before task creation so it a) stops the Task object from being created and b) gives a nicer callstack (and maybe other reasons).

The test (expecting that ArgumentNullException was thrown synchronously) is failing "with the wrong reason", because Xunit has a secondary thing that says if you call Assert.Throws with a lambda that returned an awaitable that instead of saying "it didn't throw" it tells you "maybe you meant ThrowsAsync".